### ISSUE TITLE
Implement the stack concept

### ISSUE THREAD

**Issue by lapwingcloud** [opentofu/opentofu#931]
### OpenTofu Version

```shell
OpenTofu v1.6.0-alpha5
on linux_amd64
```

### Use Cases

Stack a way to split monolith infrastructure configurations to smaller subsets, each subset of configurations manages the corresponding subset of the infrastructure, which has the following benefit 

- reduce blast radius 
- faster to apply 
- less api costs
- make collaboration easier (like less merge conflicts) 

Terraform announces they will be adding the stack in their official implementation https://www.hashicorp.com/blog/terraform-stacks-explained 

### Attempted Solutions

It already be resolved with different custom implementations, e.g. we can simply use folders as stacks, and have different variables for different environments

```
├── service-bar
│   ├── database.tf
│   ├── ec2.tf
│   ├── elb.tf
│   └── vars
│       ├── all.tfvars
│       ├── production.tfvars
│       └── staging.tfvars
└── service-foo
    ├── database.tf
    ├── ec2.tf
    └── elb.tf

```

This will usually require a custom wrapper to tofu or terraform to pass the different vars when running the plan and apply

This customization logic also implemented in several third party tools including terragrunt, terramate etc. 

### Proposal

all the third party or customized implementation works great, but since terraform is going to standardize it I'd also like to see it being standardized in tofu so we can be more consistent and focus on 1 tool (tofu) itself for scaling the configurations for large projects

**Comment by janosdebugs** [comment]
Hey @lapwingcloud thank you so much for this suggestion! We are currently working towards the 1.6 stable release. Any additional features will have to wait until that release is finished.

**Comment by gtirloni** [comment]
Maybe this can wait until Hashicorp actually delivers the feature so OpenTofu can learn what works or not with the stack and inform any technical specifications.

**Comment by skyzyx** [comment]
This sounds very much like how Terragrunt can be used. I, for one, would love to see more of Terragrunt's functionality implemented natively in OpenTofu.

**Comment by muandane** [comment]
Would love to see this in tofu instead of having to opt for a terraform/opentofu wrapper.... terraform are actually working on this but might be only for terraform Cloud using the rpcapi command

**Comment by nikolay** [comment]
Terraform Stacks is the main stopper from me fully embracing OpenTofu as Stacks really solve a lot of real-world Terraform headaches!

**Comment by EdanBrooke** [comment]
I am watching this issue closely as I, too, would like to simplify our IaC pipelines and preferably move away from Terragrunt.

We're still trying to choose between TF & Tofu. TF Stacks seems to be a HCP feature and we don't want to go down that route so I'm hopeful that Tofu can deliver in this regard.

Thank you OpenTofu maintainers for your work.

**Comment by muandane** [comment]
> I am watching this issue closely as I, too, would like to simplify our IaC pipelines and preferably move away from Terragrunt.
> 
> We're still trying to choose between TF & Tofu. TF Stacks seems to be a HCP feature and we don't want to go down that route so I'm hopeful that Tofu can deliver in this regard.
> 
> Thank you OpenTofu maintainers for your work.

I forgot where i saw this info, but stacks are not going to be ported probably to terraform outside of HCP because the Stacks feature isn't run exactly by terraform but with the automation surrounding it in HCP, so the road to Tofu stacks might be a bit further than we think since such automation would require a new mechanism to be added into the code base

**Comment by gabops** [comment]
> Nothing would prevent us from adding a stack concept to OpenTofu itself. Where I see the problem is OpenTofu's internal state. In my mind, a stack concept would involve applying one project, then applying a next project, and so on. This currently cannot be done inside a single OpenTofu instance because OpenTofu has a lot of internal state. However, if we were to get rid of the internal global state, nothing would stop us from doing something like this. As a side-effect, this may also make OpenTofu usable as a Go library with a bit of added work.
> 
> That being said, we are talking about refactoring 300k lines of code here. I would like to understand why Terragrunt is not suitable as a tool before jumping into any potential discussions about the implementation. As someone who never used Terragrunt, it seems to get the job done and simply merging that functionality into OpenTofu just for getting rid of an extra tool to use seems like a poor use of resources.

The three reasons that come to my head for preferring to get rid of Terragrunt are:

1. It forces you to pollute your project with a bunch of terragrunt.hcl files spread on a directory tree, forcing you to adopt certain practices (e.g project layout) to a point where one ends up managing a Terragrunt project rather than just a Terraform/OpenTofu project that uses a 3rd party runner.

2. It's a 3rd party dependency which is an extra moving part with its own complexity and shenanigans. Also, another element to master and to teach to someone who joins the team.

3. In big and complex projects, Terragrunt is a PITA to debug and quite slow.

Also, for what I've read in the documentation, I personally find Terraform's approach way better thought and simpler (just a bunch of files in your project's root directory: components.tfstack.hcl, deployments.tfdeploy.hcl etc), although to be fair, I haven't seen a complex use case and as someone mentioned earlier this functionality seems to rely on HCP and not in Terraform CLI itself.

**Comment by gabops** [comment]
> I wonder if this would be something to take up with the Terragrunt folks first and at the very least discuss how some of the frustrations can be addressed. They are an open source project and support OpenTofu. Us reimplementing a large part of what they are doing really seems like a wasted opportunity.

I guess that if implementing that functionality is that hard, we don't have other choice although it's impossible for the Terragrunt team to drastically change their implementation design.

Another option would be that the OpenTofu project could built its own tool (with casinos, and hookers!) that implements Terraform Stacks and call it via subcommand? This introduces a new tool which contradicts the second point I made earlier but we would still get a simpler and more cohesive tool. I don't know. Just throwing out ideas.

**Comment by yhakbar** [comment]
Hey folks!

Terragrunt maintainer here.

First, I want to say that I want to be mindful of taking up too much space in OpenTofu forums talking about Terragrunt. The OpenTofu team has quite a lot on their plates already, and worrying about how people feel about third party tools can take up a lot of headspace! We've recently been putting a lot of effort into making sure that there's better forums for the Terragrunt community to discuss and learn from each other, so if you want to discuss things that are explicitly Terragrunt, please feel free to visit one of these two and get a conversation going there:

- [Terragrunt Discord](https://discord.gg/SPu4Degs5f)
- [Terragrunt GitHub Discussions](https://github.com/gruntwork-io/terragrunt/discussions)

We know that we could have been doing a better job at that historically, and we're trying to step up our efforts at fostering a healthy community.

Second, I want to make it explicit that Gruntwork, and particularly the maintainers of Terragrunt, are strong OpenTofu advocates and want it to succeed. We want it to be the best tool possible, and meet directly with OpenTofu maintainers to try to improve OpenTofu, even when features we are pushing for might make Terragrunt less necessary for certain users. A specific example of this is how we're coordinating with the OpenTofu team on #1483, which makes [Terragrunt Provider Caching](https://terragrunt.gruntwork.io/docs/features/provider-cache/) less necessary. We're not trying to push Terragrunt use because it _fills in gaps_ that OpenTofu could sensibly provide on its own, but because it _adds_ capabilities on top of OpenTofu that improves the experience of managing IaC at scale.

All that being said, I personally think OpenTofu taking on Stacks wouldn't be the best use of resources for the OpenTofu team. There's plenty for OpenTofu to address in the way OpenTofu works today that doesn't require building out an extra tool for its own implementation of Stacks, orchestrating OpenTofu executions across multiple pieces of state. Terragrunt users have been getting that for free for a long time.

The [Terragrunt Stacks RFC](https://github.com/gruntwork-io/terragrunt/issues/3313) is designed to address that drawback of `terragrunt.hcl` file sprawl, while keeping the tool backwards compatible for users that already have Terragrunt configurations they don't want to rewrite. It's not a new binary, or a dependency on some proprietary service or anything. It's just a way of representing existing Terragrunt configurations more succinctly using the same open source tool. I wrote a [long walkthrough](https://github.com/yhakbar/terragrunt-3313-stacks-walkthrough) that breaks this down, and has a simple bash script to mock out the behavior so that you can get a feel for what Terragrunt Stacks will be like. We also have more planned for improving the ergonomics of Terragrunt Stacks and the overall experience of using Terragrunt to orchestrate OpenTofu, and are always looking for feedback.

If you don't mind a lot of `terragrunt.hcl` files, however, most of what I've heard regarding how Terraform Stacks work (I haven't tried it, though, and don't know anybody in real life that has, TBH) can be achieved using Terragrunt today. Gruntwork has customers successfully managing massive IaC estates using Terragrunt and OpenTofu without any notion of Stacks.

There's a [lot planned for Terragrunt 1.0](https://blog.gruntwork.io/the-road-to-terragrunt-1-0-4d5a0b416086), and if Terragrunt users feel like it doesn't meet their needs today or as described in that blog post, we want to hear about it. The reason we aren't announcing a release date is that we want to give time to change things, as a result of community feedback, before we land on a stable interface for a 1.0 that we plan on supporting for a _long_ time with no breaking changes. We're also going to be building things out in the open and gradually, so users won't suddenly manage their infrastructure differently all at once after a specific 1.0 release.

Hopefully this gives some insight as to how we're seeing this issue on the Terragrunt side!

**Comment by nikolay** [comment]
@yhakbar Any rough idea when v1 would be out? "A lot planned" sounds like "many months" if not "many years" in the future.

**Comment by yhakbar** [comment]
@nikolay No projected timeline right now. Definitely not "many years", though. 

If you have any concerns preventing you from adopting Terragrunt today, you don't have to wait for 1.0! Share your feedback on Terragrunt forums mentioned above, and we'll take your feedback seriously.

**Comment by nikolay** [comment]
@yhakbar So, over one year for sure. Thanks!

**Comment by gabops** [comment]
In an attempt to try to keep this conversation fruitful I have to say that, despite Terragrunt aspirations and promises (which, don't get me wrong, I'm grateful for), I still think that is a good idea that OpenTofu should provide a native way of handling stacks (or at least, a way of controlling the order and parallelization of "backend grouped resources", call it "stacks" or whatever). Relying on a 3rd party tool for such a common pattern on medium/large serious projects feels wrong. I think we have a great opportunity here for improving the usage of OpenTofu.

**Comment by marziply** [comment]
@gabops I completely agree. With all due respect to the Terragrunt team, I don't want to depend on third party tooling for a feature that I believe should definitely be native to OpenTofu.

**Comment by nikolay** [comment]
@abstractionfactory Let's not go a wild goose chase, please! I asked for a timeframe, and based on the answer, it seems that v1 is at least a year in the future and many of us need a solution soon. Similarly to @gabops and @marziply, I never want to depend on third-party tools (Terragrunt, Terramate, etc.), which also use an approach that I believe is a patchwork around HashiCorp's Terraform limitations and years of unwillingness to fix them. Many of the problems Terragrunt "solves" will be soon solved by OpenTofu, for example, dynamic providers, allowing variables and locals everywhere, etc. So, honestly, it's better to focus all efforts on solving the long-standing Terraform issues with OpenTofu rather than relying on third-party tools providing patchworks, which tools also don't seem to have the same dedicated resources and community to move fast and deliver something much better than Terraform Stacks, which also seems to be a separate tool, which they purposely kept separate to lure customers into their Captain Cook's RUM-infused product and not solve the Terraform issues so people are pressed to pay for Terraform Cloud. It could seem like a conspiracy theory, but it's more like Occam's Razor given the switch to BUSL and the acquisition by IBM.

**Comment by DavidGamba** [comment]
Implementing a thin wrapper around the existing terraform/tofu binary to implement the stacks feature is "fairly" easy but as mentioned before it requires multiple invocations of the command.

There a few things that stand in the way of a "very" easy thin wrapper implementation, which maybe should be fixed first, so that it becomes trivial.

1. Running multiple instances of Terraform init causes race conditions when writing to the plugin cache, so those init calls need to be run in serial for each component in the stack. The benefit of the plugin cache is that if a provider is found it will not re-download it unlike the `terraform provider mirror` command. 

Once you have your providers in the cache I have had no issues running init in parallel otherwise. In CI I only rebuild that cache when I see changes to the lock files.

2. Running two workspaces of the same component in parallel results in a race condition to the `.terraform/terraform.tfstate` file.
When running stacks, I have to make sure to use a different `TF_DATA_DIR` for each workpace.

3. Obviously use the `TF_WORKSPACE` env var rather than doing a `terraform workspace select`.

**Comment by nikolay** [comment]
@abstractionfactory I will try to distill and explain the challenges and how I worked around them with pure Terraform and HPC Terraform. At one point I was thinking that all this could be handled by a "workflow" provider, which runs Terraform as a provider within the provider. Crazy, right?

**Comment by nikolay** [comment]
@abstractionfactory I will try to distill and explain the challenges and how I worked around them with pure Terraform and HPC Terraform. At one point I was thinking that maybe all this could be handled by a "workflow" provider, which runs Terraform as a provider within the provider. Crazy, right?

**Comment by nikolay** [comment]
@abstractionfactory Both are great, but taking the Terragrunt approach in the OpenTofu Pipes is a bit dangerous unless there's a state lock for the inputs until the dependent process is done. Otherwise, you may get inconsistent results. I didn't know how dependencies in Terragrunt work and I thought that it just generates a bunch of `terraform_remote_state` obejcts but it does what the Pipes proposal does, and I got disappointed a little bit.

Anyway, this would work, too, but I would suggest adding something simple to OpenTofu - types and the ability to type outputs as well. IRL, outputs could produce really complex data structures, so, it was fine to have dynamic typing before Stacks/Pipes, but with that in mind, I think outputs should be typed, too, and types should be reused, but that could be accomplished by symlinking for starters. So, something like `types.tofu` could be used to type the outputs and the variables (inputs) of the consuming workspace via, initially, a symlink. Later, there could be an import/include feature.

One other reason for the need of types is also for the defaults they carry. One big issue with Terraform Cloud (no, I won't call it HCP Terraform) is that if you have `terraform_remote_state` and that state doesn't exist, you get an error - even if you have defaults specified! This forces you to modify the code just to make interconnected things work! With types, you don't have to repeat all this as the types will have `optional()` with defaults or just default value overall. Types also could be nested in another types - right now, often I have to repeat the same type definition multiple times!

Also, with `terraform_remote_state`, you can also do `for_each` - with fixed inputs, this can't be done.

The refactor feature is great as it really requires skills to extract resources from state although recently, I started to use imports and `state rm`'s.

**Comment by nikolay** [comment]
@abstractionfactory I added a types proposal: https://github.com/opentofu/opentofu/issues/2230

**Comment by daviewales** [comment]
I currently have a small repository structured like this:

```
.
├── global
│   └── storage
│       └── remote-state
└── prod
    └── compute
        ├── example-vm
        └── other-example-vm
```

The subdirectories in the final level are each independent TF configurations, so can each be deployed independently from the others.
I'm the only person at my organisation who uses OpenTofu, so I'm trying to keep the process as simple as possible for the next person. For this reason, I'm trying to avoid the complexity of adding an additional tool such as Terragrunt, to reduce the barrier to entry for my successor. However, there are a number of Terragrunt features which would make this easier to manage. However, as I don't need _all_ the Terragrunt features, and have a very small project, it would be nice if some of these features were available natively in OpenTofu:

## Reusable backend configuration

Terragrunt makes it easy to [define the backend configuration in one place](https://terragrunt.gruntwork.io/docs/getting-started/overview/), and re-use it everywhere, including automatically setting the state key from the directory path. In Terragrunt, this looks like defining a top level `root.hcl` file with the following configuration (example from linked Terragrunt docs):

``` hcl
remote_state {
  backend = "s3"

  generate = {
    path      = "backend.tf"
    if_exists = "overwrite_terragrunt"
  }

  config = {
    bucket = "my-tofu-state"

    key            = "${path_relative_to_include()}/tofu.tfstate" # <--
    region         = "us-east-1"
    encrypt        = true
    dynamodb_table = "my-lock-table"
  }
}
```

Then, the individual units include this configuration as follows, which automatically generates a key from the path:

``` hcl
include "root" {
  path = find_in_parent_folders("root.hcl")
}
```

## Deploy multiple units

In general, I want to deploy each unit independently, as usually I will only change one at a time.
However, occasionally I make a change across the stack.
In this case, it would be nice to have the option to run `tofu apply` for everything.
Terragrunt supports this with `terragrunt run-all apply`.

## Why not just use Terragrunt then?

1. For the sake of simplicity, I want it to remain possible to run `tofu apply` directly in each unit. I don't want my successor to need to learn two separate tools to get started.
2. I don't want the configuration logic split across Tofu and Terragrunt files, as this makes Terragrunt required, rather than an optional enhancement.
3. I don't want all those `terragrunt.hcl` files.

**Comment by yhakbar** [comment]
Hey @daviewales ,

Terragrunt maintainer here.

This is not to dissuade any particular course of action on the OpenTofu side, but just wanted to share some news about advancements on the Terragrunt side. If you aren't aware, we're about to mark the [experimental stacks feature](https://terragrunt.gruntwork.io/docs/reference/experiments/#stacks) ([#3313](https://github.com/gruntwork-io/terragrunt/issues/3313) as generally available. We also have a [migration guide](https://terragrunt.gruntwork.io/docs/migrate/terragrunt-stacks/) for any users that are currently using Terragrunt without Terragrunt Stacks.

The short explanation of what Terragrunt Stacks do is that they allow users to author a single `terragrunt.stack.hcl` file instead of committing a bunch of `terragrunt.hcl` files to their repositories to achieve the same benefits (configurations that are DRY (Don't Repeat Yourself), low blast radius and convenient to collaborate on with multiple engineers, regardless of IaC scale).

In general, Terragrunt has been designed to allow users to opt in to the features they want, based on real world use-cases, so if you do decide to give Terragrunt another shot (or just want another community to join with a bunch of IaC practitioners), I'd encourage you to join the [Terragrunt Discord](https://discord.gg/YENaT9h8jh), and chat with me about how you're managing your infrastructure. My team and I would love to help how we can.

**Comment by daviewales** [comment]
Thanks @yhakbar,

Terragrunt Stacks look like a good step forward for Terragrunt, although I think they are probably more suited to bigger deployments than what I'm working with currently. I'd certainly look into it if I needed to reuse modules across multiple deployments, or within larger deployments. But I don't think it quite fits my small niche use case.

**Comment by apparentlymart** [comment]
Hi all,

I just wanted to note, since there were some comments in the meantime asking what's going on with this, that over in Terraform land the "Stacks" concept seems to remain purely an HCP Terraform feature that's coupled with that commercial product, rather than something that can be used in isolation using only Terraform Community Edition. If that changes then of course we will think about what that could change for OpenTofu, but for now this issue is essentially still "on hold".

However, we are separately talking about the more general idea of making OpenTofu work better for larger configurations (which some like to call "Terraliths") so that it might become less important to decompose configurations into smaller units in the first place, because OpenTofu _itself_ would offer features like planning and applying changes to only parts of the system at a time. Perhaps a shorthand (but imprecise) way to think about that approach is "what if normal module calls could do some or all of the things that _Stack Components_ or _Terragrunt Units_ can do, without introducing a new concept?".

That discussion is still very early and in the "information gathering" phase, but if you want to follow along or participate then https://github.com/opentofu/opentofu/issues/2860 is currently acting as our main hub for that work.

Thanks!

**Comment by ImIOImI** [comment]
I had an interesting discussion about stacks and exposing outputs today. One of the biggest pain points, imo, using stacks and sharing the outputs is the lack of any kind of autocomplete functionality. If there was a protocol for a theoretical server with an API that exposes stack outputs from a state file, such things could be queryable and added into tofu-ls... and that would be super slick.

**Comment by tmccombs** [comment]
Another pain point I've encountered with respect to outputs and referencing remote state that may be relevant to this is that using the outputs from another state requires reading the entire state file, which can lead to significant memory and cpu usage when reading the depency state files are large and/or numerous.  Having a more general way to consume just the outputs could mitigate that.

**Comment by ImIOImI** [comment]
@tmccombs I made [an issue a long time ago](https://github.com/opentofu/opentofu/issues/949) about a re-imagining of how to share outputs from one project to another and it has received precisely 0 votes. So, maybe if you want to rethink things avoid this clearly terrible idea? 

<img src="https://media0.giphy.com/media/v1.Y2lkPWJkM2VhNTdlZ2RlYWV5NjgyejhyOTFhYjloOTZ6MnJxaDRxZnczOWZsZXZvdDN2NyZlcD12MV9naWZzX3NlYXJjaCZjdD1n/bWooWMiwxbiiA/giphy.gif"/>

I still think its a good idea. Creating some kind of interface object to specifically hold the contract between producer and consumer that publishes the "public" outputs of one stack to another would help me tremendously.

**Comment by DavidGamba** [comment]
On the issue of outputs, I have never advocated for using the state file as a means to connecting two components (two different terraform projects belonging to a stack). 
The problem is that a custom json in an S3 bucket is not only not the fastest way to share data but also a Terraform only one.

My outputs normally go to a tool like AWS Parameter Store or AWS Secrets Manager (depending on their nature). In that way I can not only query them in a separate component with a data block but I also get to leverage that data in other tooling like a script or a separate technology.

**Comment by ImIOImI** [comment]
@DavidGamba this is a great point... 
```
The problem is that a custom json in an S3 bucket is not only not the fastest way to share data but also a Terraform only one.
```

...and also one of the reasons why sharing outputs can be non-trivial in contexts that aren't Tofu specific. For simple strings and such, using parameter store works a charm, but for complex objects there isn't a good way to serialize and unserialize the data into Tofu types. Even storing it in JSON or yaml isn't type safe.

**Comment by red8888** [comment]
> rtant to decompose configurations into smaller units in the first place, because OpenTofu _itself_ would offer features like planning and applying changes to only parts of the system at a time. Perhaps a shorthand (but imprecise) way to think about that approa

I was just going open a feature request asking if you guys were planning to support ordering/dependency chaining of providers-  is that kind of what your talking about here?

### Specific example
I want to have a single terraform module that creates a GKE cluster and connects to that cluster after its created to apply kubernetes_manifest resources. I can't do that currently because the cluster has to exist before the k8s provider can connect to it and terraform evals providers immediately.

For my little modules stacks seem like overkill. All I really need is for terraform to delay evaluation of the k8s provider till the cluster is created. Could this be a builtin feature of opentofu sans stacks?

**Comment by apparentlymart** [comment]
Hi @red8888,

I think https://github.com/opentofu/opentofu/issues/1685 is a more specific issue representing the problem you've described.

And indeed, I think that's a problem worth solving regardless of whether OpenTofu gets any feature called "stacks". Unfortunately when we tried an initial experiment with improving on that we found that the `hashicorp/kubernetes` provider doesn't produce a trustworthy signal for "my configuration is too unknown to be able to initialize on this round" -- it reports that even in some situations where provider initialization _could_ work today -- and so we had to revert it as described in https://github.com/opentofu/opentofu/issues/1685#issuecomment-3672725880.

I'm hoping we can try that again later in the work on https://github.com/opentofu/opentofu/issues/3414, but we do still need to figure out what to do to avoid the provider's unreliable signal making things worse for configurations that are already working today. We'll discuss that more over in https://github.com/opentofu/opentofu/issues/1685 once we get there.

**Comment by DavidGamba** [comment]
@red8888 you can definitely hack Terraform to create a cluster and connect to it in the same component [1]. However, because it is a hack you have to deal with the consequences at destroy time and do it in multiple stages [2].
Going outside of Terraform to orchestrate the execution of a few components is more maintainable IMHO so you don't have to carry your provider inception around.

[1] https://github.com/hashicorp/terraform/issues/24476#issuecomment-1178213409
[2] Funny thing, is that even without hacks, there are many resources that require 2 stage deletions. Like removing an RDS db deletion protection in one step and the deleting it in the next.

**Comment by Clockwork-Muse** [comment]
@red8888 - To echo @DavidGamba , the more "usual" way to do something like this is to set up a GitOps provider like Flux or ArgoCD, and point them at a git repo.
There are multiple benefits to doing this, including dealing with config drift, and not relying on explicit push workflows (and if you take the time, automated ephemeral PR-time deployments for testing).

**Comment by ImIOImI** [comment]
@red8888 Stacks are definitely the best way to manage k8s cluster creation and provisioning currently. However, It would indeed be very useful to be able to package cluster creation and table stakes provisioning together. One of the other benefits is that it decouples the machine it runs on and the software that its running, so you only have to delete the cluster then clear the state associated with provisioning stacks as opposed to how TF would naturally attempt to uninstall all the charts THEN destroy the cluster. Imagine the destroy timing out on a namespace finalizer...

@apparentlymart if we figure out how to couple provisioned hardware with the software it runs into a single stack, it'll increase the need for lifecycle rules on said software as captured in https://github.com/opentofu/opentofu/issues/3148
